edc-extensions/hashicorp-vault/src/main/java/org/eclipse/dataspaceconnector/core/security/hashicorp/HashicorpVaultClient.java

```diff
@@ -0,0 +1,205 @@
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import java.io.IOException;
+import java.net.URLEncoder;
+import java.nio.charset.StandardCharsets;
+import java.util.HashMap;
+import java.util.Objects;
+import java.util.concurrent.CompletableFuture;
+import lombok.RequiredArgsConstructor;
+import okhttp3.Call;
+import okhttp3.Callback;
+import okhttp3.MediaType;
+import okhttp3.OkHttpClient;
+import okhttp3.Request;
+import okhttp3.RequestBody;
+import okhttp3.Response;
+import org.eclipse.dataspaceconnector.spi.EdcException;
+import org.eclipse.dataspaceconnector.spi.result.Result;
+import org.jetbrains.annotations.NotNull;
+
+@RequiredArgsConstructor
+class HashicorpVaultClient {
+  private static final String VAULT_TOKEN_HEADER = "X-Vault-Token";
+  static final String SECRET_KEY = "key";
+  private static final String VAULT_SECRET_DATA_PATH = "/v1/secret/data/";
+  private static final String VAULT_SECRET_METADATA_PATH = "/v1/secret/metadata/";
+
+  private final HashicorpVaultClientConfig config;
+  private final OkHttpClient okHttpClient;
+  private final ObjectMapper objectMapper;
+
+  private String getSecretDataUrl() {
+    return String.format("%s%s", config.getVaultUrl(), VAULT_SECRET_DATA_PATH);
+  }
+
+  private String getSecretMetadataUrl() {
+    return String.format("%s%s", config.getVaultUrl(), VAULT_SECRET_METADATA_PATH);
+  }
+
+  @NotNull
+  CompletableFuture<Result<String>> getSecretValue(String key) {
+    Request request =
+        new Request.Builder()
+            .url(
+                String.format(
+                    "%s%s", getSecretDataUrl(), URLEncoder.encode(key, StandardCharsets.UTF_8)))
+            .get()
+            .header(VAULT_TOKEN_HEADER, config.getVaultToken())
+            .build();
+
+    CompletableFuture<Result<String>> future = new CompletableFuture<>();
+    okHttpClient
+        .newCall(request)
+        .enqueue(
+            new Callback() {
+              @Override
+              public void onResponse(@NotNull Call call, @NotNull Response response) {
+                if (response.code() == 404) {
+                  future.complete(null);
+                  return;
+                }
+                if (response.isSuccessful()) {
+                  try (var body = response.body()) {
+                    if (body == null) {
+                      future.completeExceptionally(
+                          new EdcException("Received an empty body response from vault"));
+                    } else {
+                      GetHashicorpVaultEntryResponsePayload payload =
+                          objectMapper.readValue(
+                              body.string(), GetHashicorpVaultEntryResponsePayload.class);
+                      String value =
+                          Objects.requireNonNull(payload.getData().getData().get(SECRET_KEY));
+                      future.complete(Result.success(value));
+                    }
```

Comment:
You can get rid of this else if you return in the if body before:
```suggestion
                      return;
                    }
                    
                    GetHashicorpVaultEntryResponsePayload payload =
                        objectMapper.readValue(
                            body.string(), GetHashicorpVaultEntryResponsePayload.class);
                    String value =
                        Objects.requireNonNull(payload.getData().getData().get(SECRET_KEY));
                    future.complete(Result.success(value));
```